android-components/components/browser/engine-gecko/src/main/java/mozilla/components/browser/engine/gecko/GeckoResult.kt

```diff
@@ -43,10 +44,15 @@ fun <T> GeckoResult<T>.asCancellableOperation(): CancellableOperation {
             val result = CompletableDeferred<Boolean>()
             geckoResult.cancel().then(
                 {
+                    // Notify the geckoResult that there were able to cancel the request
+                    if (it == null || it == false) {
+                        geckoResult.completeExceptionally(Exception("Unable to cancel"))
+                    }
```

Comment:
Before writing tests, I would like to share a possible solution. 

The main issue that we seeing in [bug 1868249](https://bugzilla.mozilla.org/show_bug.cgi?id=1868249), it's the user is trying to cancel an installation request, but it fails as the file was already download and we are far away to the point on cancelling, we already downloaded the `xpi` and showed the permission prompt . 

On the UI, we see the overlay not hiding as a result of https://github.com/mozilla-mobile/firefox-android/pull/4735/files#r1425593558.


Also after the user declines the permissions, clicks on an details page and goes back (This can be watch on the reporter video), we are stuck loading, as we didn't get any response from Gecko, presumably because of failed [cancellation request ](https://searchfox.org/mozilla-central/source/toolkit/mozapps/extensions/internal/XPIInstall.jsm#1442). 

Added some logs to the [cancellation code path](https://searchfox.org/mozilla-central/source/mobile/android/modules/geckoview/GeckoViewWebExtension.sys.mjs#428).

```
2023-12-13 09:59:19.938 25465-25570 GeckoViewConsole        org.mozilla.fenix.debug              D  Unable to cancel installId ad713f81-fbe1-4d17-bb1a-8d46e3349887 error Error: Cannot cancel install of https://addons.mozilla.org/android/downloads/file/4206186/noscript-11.4.29.xpi from this state (1)(resource://gre/modules/addons/XPIInstall.jsm:1441:15) JS Stack trace: cancel@XPIInstall.jsm:1441:15
```

Reply:
This possible solution when we notice that we are not able to cancel the installation request, we proceed to let the caller know so it could clean up properly. Another further investigation is to research why gecko is not notifying us back a [failed cancellation request](https://searchfox.org/mozilla-central/source/toolkit/mozapps/extensions/internal/XPIInstall.jsm#1442) happens .